pkg/controller/garbagecollector/graph_builder.go

```diff
@@ -581,12 +581,14 @@ func (gb *GraphBuilder) addUnblockedOwnersToDeleteQueue(removed []metav1.OwnerRe
 }
 
 func (gb *GraphBuilder) processTransitions(oldObj interface{}, newAccessor metav1.Object, n *node) {
-	if startsWaitingForDependentsOrphaned(oldObj, newAccessor) {
+	if startsWaitingForDependentsOrphaned(oldObj, newAccessor) ||
+		(n.isDeletingDependents() == false && beingDeleted(newAccessor) && hasOrphanFinalizer(newAccessor)) {
 		klog.V(5).Infof("add %s to the attemptToOrphan", n.identity)
 		gb.attemptToOrphan.Add(n)
 		return
 	}
-	if startsWaitingForDependentsDeleted(oldObj, newAccessor) {
+	if startsWaitingForDependentsDeleted(oldObj, newAccessor) ||
+		(n.isDeletingDependents() == false && beingDeleted(newAccessor) && hasDeleteDependentsFinalizer(newAccessor)) {
```

Comment:
I am not sure this change would result in any different behavior. `beingDeleted(newAccessor) && hasDeleteDependentsFinalizer(newAccessor)` is checked inside `startsWaitingForDependentsDeleted`. and `n.isDeletingDependents()` should always be false because the only place that marks deletingDependents is down below. Could you elaborate a bit more?

Reply:
@jiahuif 
 i add a judge that (n.isDeletingDependents() == false && beingDeleted(newAccessor) && hasDeleteDependent)

it will be true only when gc lost the event (old object without deleteTimestamp and foregroundDeletion, new object with deleteTimestamp and foregroundDeletion), but **due to resyn of infomer or user's retry** , gc will watch the next update event
(old object with deleteTimestamp and foregroundDeletion, new object with deleteTimestamp and foregroundDeletion)

and gc will modity rs.deletingDependents is true。and the process will not be stuck


see more information in this issue
https://github.com/kubernetes/kubernetes/issues/115570#issuecomment-1425103268